### PR TITLE
Correction of thresholding

### PR DESCRIPTION
**Pull request by nkt1546789**
We should threshold the probability instead of its label.

### CODE REVIEW

**Comment by matt-peters** [comment]
Ah you are right we should threshold based on probability not label.  Using the `predict` method on the block model is a holdover from prior code that didn't use scikit-learn.

However, this PR introduces a few other problems that also need to be addressed before merging:

* modifying the `block_model` interface to use `predict_proba` instead of `predict` breaks some existing functionality and tests
* The classifiers in scikit-learn don't have a deterministic ordering of class labels so you need to inspect the `classes_` attribute on a trained model to determine which index has label 1.  In your code you have hard coded this to the second index, but in the future this many very well change to the first index.  We need to future proof this.

The easiest solution I can see is to leave the `block_model` interface as is and introduce a `SklearnWrapper` class that wraps implements it.  It needs a `predict` method that returns probability of class == 1 using a trained scikit-learn model via `predict_proba` and `classes_`.

**Comment by nkt1546789** [comment]
Thanks for your feedback. I learned a lot and got your points. 
For now, I'm using the code below which works fine. 
```python
from dragnet import content_extractor

classes=list(content_extractor._block_model.classes_)
positive_idx=classes.index(1)
def extract(html,block=False,threshold=0.5):
    features,blocks=content_extractor.make_features(html)
    scores=content_extractor._block_model.predict_proba(features)[:,positive_idx]
    if block:
        return [block for i,block in enumerate(blocks) if scores[i]>=threshold]
    return " ".join([block.text for i,block in enumerate(blocks) if scores[i]>=threshold])
```
But I'm don't know how to incorporate this kind of modification into your code. 
Do you mean that we should make SklearnWrapper class by extending ContentExtractionModel class?

**Comment by matt-peters** [comment]
Your code snippet looks good and includes much of what you need.  The `ContentExtractionModel` class glues together a few different sub-components and processes an entire document end to end. One of these is the machine learning `block_model`.  I'm suggesting to implement a simple wrapper that will be substituted for the block model, something like:

```python
class SklearnWrapper(object):
    def __init__(self, skmodel):
        # skmodel implements predict_proba and has classes_ attribute
        self._skmodel = sklmodel
        classes = list(skmodel.classes_)
        self._positive_idx = classes.index(1)

    def predict(self, X):
        return self._skmodel.predict_proba(X)[:, self._positive_idx]
```

Then in `models.py` when the pre-trained models are loaded you can monkey patch the existing `_block_model` with instances of this class.  Ideally we wouldn't need this last step as it's kinda a hack but it is simpler and lighter weight then replacing the existing pickle files (and easy to clean up if the models are ever retrained).

**Comment by nkt1546789** [comment]
OK. I will implement it. So SklearnWrapper class only used for prediction?

**Comment by nkt1546789** [comment]
I have a suggestion although this is not related to this PR. 

I think the threshold affects the performance, e.g. evaluation score, so why don't you choose the threshold using cross-validation. This approach may produce better results. 

If I remember correctly, this approach (choosing the threshold of conditional probability) is called cost-sensitive learning and most famous paper is "The foundation of cost-sensitive learning" by Elkan. Maybe you can refer it.

**Comment by matt-peters** [comment]
@nkt1546789 Sorry didn't see that you had pushed some more commits. The sklearn wrapper looks good, except can you remove the binary `dist/dragnet-1.0.1-py2.7-macosx-10.10-intel.egg`.  Then I'll merge.

**Comment by matt-peters** [comment]
Agree that setting the threshold should be a part of cross validation.  I did this for the original trained model that used the `LogisticRegression` class in `mozsci` (that has a different interface then sklearn with probability vs label prediction).  I didn't do it during the last iteration where we switched to sklearn's random forest classifier otherwise I'd have caught this bug.

**Comment by nkt1546789** [comment]
I have removed the file.

**Comment by nkt1546789** [comment]
Thanks a lot for the merge, great tips to write codes, and great work!
